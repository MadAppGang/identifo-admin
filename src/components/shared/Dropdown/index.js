import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Dropdown extends Component {
  constructor() {
    super();

    this.state = {
      open: false,
    };

    this.open = this.open.bind(this);
    this.close = this.close.bind(this);
    this.handleGlobalClick = this.handleGlobalClick.bind(this);
  }

  componentDidMount() {
    document.addEventListener('click', this.handleGlobalClick);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleGlobalClick);
  }

  handleGlobalClick({ target }) {
    if (this.ref.contains(target)) {
      return;
    }

    this.close();
  }

  open() {
    this.setState({ open: true });
  }

  close() {
    this.setState({ open: false });
  }

  render() {
    const { children } = this.props;

    if (!children) {
      return null;
    }

    return (
      <div ref={el => this.ref = el}>
        {children({
          open: this.open,
          close: this.close,
          isOpen: this.state.open,
        })}
      </div>
    );
  }
}

Dropdown.propTypes = {
  children: PropTypes.func,
};

Dropdown.defaultProps = {
  children: null,
};

export default Dropdown;
