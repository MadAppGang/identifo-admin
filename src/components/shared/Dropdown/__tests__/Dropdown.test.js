import Dropdown from '..';

describe('<Dropdown />', () => {
  test('renders nothing if children function was not provided', () => {
    expect(shallow(<Dropdown />)).toMatchSnapshot();
  });

  test('calls children function on render', () => {
    const render = jest.fn();
    shallow(<Dropdown>{render}</Dropdown>);
    expect(render).toHaveBeenCalled();
  });

  test('renders what children function has returned', () => {
    const render = jest.fn(() => (
      <div>Hello, world!</div>
    ));
    const output = shallow(<Dropdown>{render}</Dropdown>);

    expect(output.contains(<div>Hello, world!</div>)).toBe(true);
  });

  test('calls passed in render function with isOpen=false by default', () => {
    const render = jest.fn(() => null);

    shallow(<Dropdown>{render}</Dropdown>);
    expect(render.mock.calls[0][0].isOpen).toBe(false);
  });

  test('calls passed in render function with isOpen=true after "open" prop was called', () => {
    const render = jest.fn(({ open }) => (
      <button type="button" onClick={open}>Click me</button>
    ));

    const output = shallow(<Dropdown>{render}</Dropdown>);

    output.find('button').simulate('click');

    const secondCall = render.mock.calls[1];

    expect(secondCall[0].isOpen).toBe(true);
  });
});
