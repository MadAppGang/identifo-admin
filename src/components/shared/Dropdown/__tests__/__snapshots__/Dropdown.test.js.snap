// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders nothing if children function was not provided 1`] = `""`;
